### DISCUSSION TITLE
Typescript: How to say the type of PrismaClient() when defining readReplicas

### DISCUSSION THREAD

**Discussion by mkevison** [prisma/prisma#23035] [Q&A] [answered]
I have a multi-tenant system with each tenant having its own database, each one has a write and a read database

If I didn't use read replicas it would be easily implemented this way:
```ts
class PrismaClientManager{
  // the client instances cache object
  private clients: { [key: string]: PrismaClient } = {};

  // retrieve and return the tenant ID from the request object,
  getTenantId(request: Request): string {
    ...
  }

  getClient(request: Request): PrismaClient {
    const tenantId = this.getTenantId(request);
    let client = this.clients[tenantId];
    // create and cache a new client when needed
    if (!client) {
      const databaseUrl = process.env.DATABASE_URL!.replace('public', tenantId);

      client = new PrismaClient({
        datasources: {
          db: {
            url: databaseUrl,
          },
        },
      });

      this.clients[tenantId] = client;
    }

    return client;
  }
}

export class MyService {
  private prisma: PrismaClient;

  constructor(
    @Inject(REQUEST) request: Request,
    prismaClientManager: PrismaClientManager,
  ) {
    this.prisma = prismaClientManager.getClient(request);
  }

  findAll() {
    // just use this.prisma to access the database
  }
}
```
But when adding read replica:
```
.$extends(
  readReplicas({
    url: Env.get('PG_SLAVE_HOST'),
  })
```

I can't tell typescript the correct PrismaClient type

**Comment by mkevison** [comment]
```
type PrismaReplica = PrismaClient & {
  $primary: Omit<PrismaClient, '$primary' | '$replica'>
  $replica: Omit<PrismaClient, '$primary' | '$replica'>
}

let replica: PrismaReplica

replica = new PrismaClient().$extends(
  readReplicas({
    url: Env.get('REPLICA'),
  })
) as any
```
That's what I've achieved so far using any

**Comment by ludralph** [comment]
Hi @mkevison 👋 

I have shared your question internally. Either I or a team member will be able to respond to you with more details.

**Reply by ludralph** [reply, marked as answer]
Hi @mkevison, you can infer the type of an extended client as mentioned in this [section](https://www.prisma.io/docs/orm/prisma-client/client-extensions#type-of-an-extended-client) of the documentation.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by mkevison** [reply]
Thank you, that's what I needed!

##### function used exclusively to return the type
```
function getExtendedClient() {
  return new PrismaClient().$extends(
    readReplicas({
      url: Env.get('PG_SLAVE_HOST'),
    })
  )
}
```
##### Create the type based on the function
```type ExtendedPrismaClient = ReturnType<typeof getExtendedClient>```

```
let PrismaDB: ExtendedPrismaClient

PrismaDB = new PrismaClient().$extends(
  readReplicas({
    url: Env.get('PG_SLAVE_HOST'),
  })
)```
